Add C implementation for `@stdlib/math/base/special/tribonacci`

Jordan-Gallivan opened the pull request:
Resolves [#895 ](https://github.com/stdlib-js/stdlib/issues/895#issuecomment-1428981391) .

## C implementation for `@stdlib/math/base/special/tribonacci`

> What is the purpose of this pull request?

This pull request:

 - Created the following directories:

```
/stdlib/math/base/special/tribonacci/src
/stdlib/math/base/special/tribonacci/include
/stdlib/math/constants/float64/max-safe-nth-tribonacci
```

 - Created the following files:

```
/stdlib/math/base/special/tribonacci/src/main.c
tribonacci.h
max_safe_nth_tribonacci.h
```

## Related Issues

> Does this pull request have any related issues?

No.

This pull request:

-   resolves [#895 ](https://github.com/stdlib-js/stdlib/issues/895#issuecomment-1428981391) .

## Questions

> addon.c was not implemented.  Please follow up if assistance is required.


## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Jordan-Gallivan committed:
C_implementation_of_tribonacci

Created the following directories:
/stdlib/math/base/special/tribonacci/src
/stdlib/math/base/special/tribonacci/include
/stdlib/math/constants/float64/max-safe-nth-tribonacci

Created the following files:
/stdlib/math/base/special/tribonacci/src/main.c
tribonacci.h
max_safe_nth_tribonacci.h

kgryte committed:
Delete IDE file

kgryte committed:
Delete IDE file

kgryte committed:
Delete IDE file

kgryte committed:
Delete IDE file

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/include/stdlib/math/base/special/tribonacci.h:
```diff
@@ -0,0 +1,36 @@
+/**
+* Computes the nth Tribonacci number.
+*/
+double stdlib_base_tribonacci( const int32_t n);
```
```suggestion
double stdlib_base_tribonacci( const int32_t n );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/include/stdlib/math/base/special/tribonacci.h:
```diff
@@ -0,0 +1,36 @@
+#define STDLIB_MATH_BASE_SPECIAL_TRIBONACCI_H
+
+#include <stdint.h>
+#ifdef __cplusplus
```
```suggestion

#ifdef __cplusplus
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/src/main.c:
```diff
@@ -0,0 +1,103 @@
+        8607945812375585
+};
+/**
+* Computes the nth Tribonacci number.
+ *
+ * @param x     input value
+ * @return      output value
+ *
+ * @example
+ * double out = stdlib_base_tribonacci( 1);
+ * // returns 1
```
```suggestion
* Computes the nth Tribonacci number.
*
* @param n     input value
* @return      output value
*
* @example
* double out = stdlib_base_tribonacci( 1 );
* // returns 1
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/src/main.c:
```diff
@@ -0,0 +1,103 @@
+ * // returns 1
+*/
+double stdlib_base_tribonacci( const int32_t n ) {
+    if (n < 0 || n > STDLIB_CONSTANTS_FLOAT64_MAX_SAFE_NTH_TRIBONACCI ) {
```
```suggestion
    if ( n < 0 || n > STDLIB_CONSTANTS_FLOAT64_MAX_SAFE_NTH_TRIBONACCI ) {
```

kgryte commented on lib/node_modules/@stdlib/constants/float64/max-safe-nth-tribonacci/include/stdlib/constants/float64/max_safe_nth_tribonacci.h:
```diff
@@ -0,0 +1,28 @@
+/**
+* Macro for the maximum safe Tribonacci number when stored in doulbe-precision floating-point format.
+*/
+
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

kgryte commented:
@Jordan-Gallivan Looking good so far! Next, you'll want to add the following files:

```
src/addon.c
src/Makefile
binding.gyp
include.gypi
manifest.json
```

See [`@stdlib/math/base/special/fibonacci`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci/src) for examples.

kgryte commented:
You'll also want to avoid committing your IDE `.idea` files. This PR should only include additions/modifications relevant to the specific library code enhancements.

Jordan-Gallivan commented:
@kgryte my apologies for the IDE .idea file uploads.  Thanks for the feedback!

It's getting late here on the east coast, so I'll get turning on the remaining files tomorrow.

kgryte commented:
@Jordan-Gallivan No worries! Thanks for contributing!

Jordan-Gallivan committed:
Final C implementation files for tribonacci

Added the following files:
src/addon.c
src/Makefile
binding.gyp
include.gypi
manifest.json

Jordan-Gallivan committed:
Merge branch 'Jordan-Gallivan-tribonacci-Branch' of https://github.com/Jordan-Gallivan/stdlib into Jordan-Gallivan-tribonacci-Branch

Jordan-Gallivan committed:
Merge branch 'stdlib-js:develop' into Jordan-Gallivan-tribonacci-Branch

Jordan-Gallivan commented:
@kgryte just merged my branch with this pull request.  Please let me know if any changes are required on my end.

For my own edification, can you point me to some resources for better understanding the .gyp, .gypi, and manifest.json files?  I more or less just changed happy to glad in those and would like to learn about their role in this project as well as their application writ large.

kgryte commented:
@Jordan-Gallivan The `*.gyp` and `*.gypi` files are necessary for compiling a Node.js native [add-on](https://nodejs.org/api/n-api.html). For this function, we'll want the ability to compile a native add-on to allow testing the C implementation from JavaScript.

The `manifest.json` file is used to resolve C dependencies. See [`@stdlib/utils/library-manifest`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/library-manifest) for more details.

Typically, the `binding.gyp` and `include.gypi` are (more or less) the same for base mathematical functions. The `manifest.json` file varies due to build targets and/or varying dependencies.

The `addon.c` file provides the glue code to allow calling from JavaScript into C.

kgryte commented:
Thanks for adding the various files. We still have a few more steps to get this PR over the finish line. :)

Up next, would you mind adding the "native" C benchmark files? You can find reference files in [fibonacci](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci/benchmark/c/native).

kgryte commented:
You'll want both `benchmark/native/benchmark.c` and `benchmark/native/Makefile`.

Jordan-Gallivan commented:
@kgryte Thank you so much for the explanation!  This is a whole new level compared to the basics I'm familiar with.

Absolutely.  I'll get it knocked out first thing tomorrow.

Jordan-Gallivan committed:
native benchmark.c and Makefile

Added the following to the tribonacci directory:
/benchmark/c/native/benchmark.c
/benchmark/c/native/Makefile

Jordan-Gallivan committed:
Merge branch 'Jordan-Gallivan-tribonacci-Branch' of https://github.com/Jordan-Gallivan/stdlib into Jordan-Gallivan-tribonacci-Branch

Jordan-Gallivan committed:
Merge branch 'stdlib-js:develop' into Jordan-Gallivan-tribonacci-Branch

Jordan-Gallivan committed:
Merge branch 'Jordan-Gallivan-tribonacci-Branch' of https://github.com/Jordan-Gallivan/stdlib into Jordan-Gallivan-tribonacci-Branch

Jordan-Gallivan commented:
@kgryte  got the native c benchmark files added just now.

kgryte commented:
@Jordan-Gallivan Awesome! Would now mind adding an [example](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci/examples/c) and [tests](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci/test/test.native.js)?

For the [`test.native.js`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci/test/test.native.js) file, the tests should be the same as `test.js` with some logic to handle if the native Node.js add-on cannot be loaded (e.g., if the add-on failed to compile). You can see the difference by comparing `test.js` and `test.native.js` in [Fibonacci](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/fibonacci/test).

Jordan-Gallivan commented:
I'd be happy to! I may come back with a few questions but I'll definitely take a stab at it

kgryte commented:
Awesome! Thanks!

Jordan-Gallivan committed:
test.native and example.c commit

Added the following:
/examples/c/example.c
/examples/c/Makefile
/test/test.native.js

Jordan-Gallivan committed:
Merge branch 'stdlib-js:develop' into Jordan-Gallivan-tribonacci-Branch

Jordan-Gallivan commented:
@kgryte I added the requested files, but wanted to give you a heads up that I could not get the example.c to compile on my end.  The file path for the header file is showing invalid.  This may be due to a lack of understanding on how to actually compile these .c files so any help would be great.  

I'm a little nervous my copy/pasting of Makefiles may not work rit large for these.  Please let me know what I can do to clean this up.

Otherwise, the actual code in the native.test.js and the example.c should be good with respect to the tribonacci constraints.

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/example.c:
```diff
@@ -0,0 +1,31 @@
+*/
+
+#include "stdlib/math/base/special/tribonacci.h"
+#include <stdio.h>
```
```suggestion
```

This header is not used in this example. I've removed also from Fibonacci.

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/example.c:
```diff
@@ -0,0 +1,31 @@
+#include <stdint.h>
+
+int main() {
+    int32_t i;
+    double v;
```
```suggestion
    double v;
    int i;
```

This triggers a lint error at the moment due to printf.

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/example.c:
```diff
@@ -0,0 +1,31 @@
+
+#include "stdlib/math/base/special/tribonacci.h"
+#include <stdio.h>
+#include <stdint.h>
```
```suggestion
```

No longer used based on proposed changes below.

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/include.gypi:
```diff
@@ -0,0 +1,53 @@
+      '<!@(node -e "var arr = require(\'@stdlib/utils/library-manifest\')(\'./manifest.json\',{},{\'basedir\':process.cwd(),\'paths\':\'posix\'}).libpath; for ( var i = 0; i < arr.length; i++ ) { console.log( arr[ i ] ); }")',
+    ],
+  }, # end variables
+}
```
```suggestion
}

```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/manifest.json:
```diff
@@ -0,0 +1,72 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/napi/unary",
+        "@stdlib/constants/float64/max-safe-nth-tribonacci"
```
```suggestion
        "@stdlib/math/base/napi/unary"
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/manifest.json:
```diff
@@ -0,0 +1,72 @@
+      ],
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/constants/float64/max-safe-nth-tribonacci"
+      ]
```
```suggestion
      "dependencies": []
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/manifest.json:
```diff
@@ -0,0 +1,72 @@
+      ],
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/constants/float64/max-safe-nth-tribonacci"
+      ]
```
```suggestion
      "dependencies": []
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/math/base/napi/unary.h"
+
+// cppcheck-suppress shadowFunction
+STDLIB_MATH_BASE_NAPI_MODULE_I_D( stdlib_base_tribonacci )
```
```suggestion
STDLIB_MATH_BASE_NAPI_MODULE_I_D( stdlib_base_tribonacci )

```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/src/main.c:
```diff
@@ -0,0 +1,103 @@
+*/
+
+#include "stdlib/math/base/special/tribonacci.h"
+#include "stdlib/constants/float64/max_safe_nth_tribonacci.h"
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/src/main.c:
```diff
@@ -0,0 +1,103 @@
+#include "stdlib/math/base/special/tribonacci.h"
+#include "stdlib/constants/float64/max_safe_nth_tribonacci.h"
+
+static const int64_t tribonacci_value[ 64 ] = {
```
```suggestion
static const int32_t STDLIB_CONSTANTS_FLOAT64_MAX_SAFE_NTH_TRIBONACCI = 63; // TODO: consider making a package similar to `@stdlib/constants/float64/max-safe-nth-fibonacci`
static const int64_t tribonacci_value[ 64 ] = {
```

Inlining the constant, as we need to actually create a package for it in order to work similar to Fibonacci.

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/test/test.native.js:
```diff
@@ -0,0 +1,94 @@
+// FIXTURES //
+
+var TRIBONACCI = require( './../lib/tribonacci.json' );
+const PINF = require("@stdlib/constants/float64/pinf");
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/test/test.native.js:
```diff
@@ -0,0 +1,94 @@
+	var v;
+	var i;
+
+	t.strictEqual( isnan( tribonacci( -3.14 ) ), true, 'returns NaN' );
+
```
```suggestion
```

Removed as non-applicable, given that, in C, we're expecting an integer. Addressed in Fibonacci.

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/test/test.native.js:
```diff
@@ -0,0 +1,94 @@
+	t.end();
+});
+
+tape( 'if provided positive infinity, the function returns `NaN`', opts, function test( t ) {
+	var v = tribonacci( PINF );
+	t.strictEqual( isnan( v ), true, 'returns NaN when provided +infinity' );
+	t.end();
+});
+
```
```suggestion
```

Similar to Fibonacci, we can skip this test.

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/test/test.native.js:
```diff
@@ -0,0 +1,94 @@
+	t.end();
+});
+
+tape( 'if provided `NaN`, the function returns `NaN`', opts, function test( t ) {
+	var v = tribonacci( NaN );
+	t.strictEqual( isnan( v ), true, 'returns NaN when provided a NaN' );
+	t.end();
+});
+
```
```suggestion
```

Similar to Fibonacci, we can skip this test.

kgryte committed:
Apply suggestions from code review

kgryte committed:
Remove header as constant value now inlined in implementation

kgryte commented:
@Jordan-Gallivan I've touched up the PR. Two last small changes:

1. Can you set the `gypfile` setting in the `package.json`? See [fibonacci](https://github.com/stdlib-js/stdlib/blob/7167ddba742218d3d43719a276c0471bb05438f6/lib/node_modules/%40stdlib/math/base/special/fibonacci/package.json#L17).
2. Can you add a `benchmark/benchmark.native.js` file? See [fibonacci](https://github.com/stdlib-js/stdlib/blob/7167ddba742218d3d43719a276c0471bb05438f6/lib/node_modules/%40stdlib/math/base/special/fibonacci/benchmark/benchmark.native.js).

Once those changes are made, would you mind pulling down the changes and running the following commands from the root of the project:

```bash
# Compile and run the C example file:
make examples-c EXAMPLES_FILTER=".*/math/base/special/tribonacci/.*"

# Compile and run the C benchmarks:
make benchmark-c BENCHMARKS_FILTER=".*/math/base/special/tribonacci/.*"

# Compile the add-on:
make install-node-addons NODE_ADDONS_PATTERN=math/base/special/tribonacci

# Run the unit tests:
make test TESTS_FILTER=".*/math/base/special/tribonacci/.*"
```

If all goes to plan, each of the above commands should successfully run and all the unit tests pass. :)

Jordan-Gallivan commented:
@kgryte this one's definitely on me.  I should have brought up earlier that the file paths for the #includes looked off and saved you a bunch of clean up.  I'll finish this up later today and let you know if the tests pass.

kgryte commented:
No worries! Thanks for continuing to work on this!

Jordan-Gallivan committed:
Merge branch 'stdlib-js:develop' into Jordan-Gallivan-tribonacci-Branch

Jordan-Gallivan committed:
benchmark.native.js and changes to package.json

Added a benchmark.native.js file and added the gyp file settings to the package.json

Jordan-Gallivan committed:
Merge branch 'Jordan-Gallivan-tribonacci-Branch' of https://github.com/Jordan-Gallivan/stdlib into Jordan-Gallivan-tribonacci-Branch

Jordan-Gallivan commented:
@kgryte added the requested changes. 

I was unable to compile the .c files again.  I keep getting a "File not found" referencing the header files for the main.c and example.c.  As a sanity check, I tried to compile an existing project (absolute value) and popped the same errors.  I'm fairly confident it's an issue on my end but let me know if you're seeing the same.

kgryte commented:
How are you trying to compile the C files?

kgryte commented:
Make sure you are using the `make` commands documented above (see [comment](https://github.com/stdlib-js/stdlib/pull/903#issuecomment-1434068060)). If you attempt to invoke the Makefile targets directly (without setting specific environment variables), it won't work.

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonacci/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,61 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = floor( randu()*79.0 );
```
```suggestion
		x = floor( randu()*64.0 );
```

kgryte committed:
Fix value range

Jordan-Gallivan commented:
@kgryte I was compiling them via command line from the root of the project as instructed.  I'm pretty confident the issue is on my end since I can't compile abs either (with the appropriate changes to the commands).  Sorry to keep this PR going with the issues on my end but please lemme know what I can do to finish this thing up.  This comment is going to be huge, but below is the readout from the examples:

`(base) jordangallivan@Jordans-MBP-6 stdlib % make examples-c EXAMPLES_FILTER=".*/math/base/special/tribonacci/.*"


Running example: /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/example.c
Resolving package path...
Package path: /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci
Resolving package manifest...
Successfully resolved package manifest.
Resolving include directories...
node:internal/modules/cjs/loader:1050
  throw err;
  ^

Error: Cannot find module 'debug'
Require stack:
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/index.js
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/[eval]
    at Module._resolveFilename (node:internal/modules/cjs/loader:1047:15)
    at Module._load (node:internal/modules/cjs/loader:893:27)
    at Module.require (node:internal/modules/cjs/loader:1113:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js:25:14)
    at Module._compile (node:internal/modules/cjs/loader:1226:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
    at Module.load (node:internal/modules/cjs/loader:1089:32)
    at Module._load (node:internal/modules/cjs/loader:930:12)
    at Module.require (node:internal/modules/cjs/loader:1113:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js',
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/index.js',
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/[eval]'
  ]
}

Node.js v18.14.0
Resolving source files...
node:internal/modules/cjs/loader:1050
  throw err;
  ^

Error: Cannot find module 'debug'
Require stack:
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/index.js
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/[eval]
    at Module._resolveFilename (node:internal/modules/cjs/loader:1047:15)
    at Module._load (node:internal/modules/cjs/loader:893:27)
    at Module.require (node:internal/modules/cjs/loader:1113:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js:25:14)
    at Module._compile (node:internal/modules/cjs/loader:1226:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
    at Module.load (node:internal/modules/cjs/loader:1089:32)
    at Module._load (node:internal/modules/cjs/loader:930:12)
    at Module.require (node:internal/modules/cjs/loader:1113:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js',
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/index.js',
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/[eval]'
  ]
}

Node.js v18.14.0
Resolving libraries...
node:internal/modules/cjs/loader:1050
  throw err;
  ^

Error: Cannot find module 'debug'
Require stack:
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/index.js
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/[eval]
    at Module._resolveFilename (node:internal/modules/cjs/loader:1047:15)
    at Module._load (node:internal/modules/cjs/loader:893:27)
    at Module.require (node:internal/modules/cjs/loader:1113:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js:25:14)
    at Module._compile (node:internal/modules/cjs/loader:1226:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
    at Module.load (node:internal/modules/cjs/loader:1089:32)
    at Module._load (node:internal/modules/cjs/loader:930:12)
    at Module.require (node:internal/modules/cjs/loader:1113:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js',
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/index.js',
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/[eval]'
  ]
}

Node.js v18.14.0
Resolving library paths...
node:internal/modules/cjs/loader:1050
  throw err;
  ^

Error: Cannot find module 'debug'
Require stack:
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/index.js
- /Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/[eval]
    at Module._resolveFilename (node:internal/modules/cjs/loader:1047:15)
    at Module._load (node:internal/modules/cjs/loader:893:27)
    at Module.require (node:internal/modules/cjs/loader:1113:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js:25:14)
    at Module._compile (node:internal/modules/cjs/loader:1226:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
    at Module.load (node:internal/modules/cjs/loader:1089:32)
    at Module._load (node:internal/modules/cjs/loader:930:12)
    at Module.require (node:internal/modules/cjs/loader:1113:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/main.js',
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/utils/library-manifest/lib/index.js',
    '/Users/jordangallivan/gitHub/stdlib/lib/node_modules/@stdlib/math/base/special/tribonacci/examples/c/[eval]'
  ]
}

Node.js v18.14.0
Compiling example...
example.c:19:10: fatal error: 'stdlib/math/base/special/tribonacci.h' file not found
#include "stdlib/math/base/special/tribonacci.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[1]: *** [example.out] Error 1
Error when attempting to compile example.
ERROR: An error was encountered during execution.

make: *** [examples-c] Error 1`

kgryte commented:
Ah. Can you run

make install-node-modules

And try again?

kgryte commented:
From your last comment, appears that you may be missing project dependencies.

Jordan-Gallivan commented:
@kgryte that did it! 
I'll make a final commit here in a second.  Thank you so much for your patience and help working through this

Jordan-Gallivan commented:
@kgryte that did it!

All checks out good on my end!  Thank you so much for your patience working through this.  I've asked to help on the negative lucas numbers, so I'll archive the above chat on my end and (hopefully) be able to knock that one out with out nearly as much hand holding.

kgryte commented:
Awesome! And no worries! To be expected for a first PR on a project which has a steep learning curve.

I'll review one final time and merge later today. Thanks again!

kgryte reviewed (approved):
LGTM. Thanks, @Jordan-Gallivan! Congratulations on your first contribution!

kgryte commented:
We forgot to update the README for tribonacci in this PR. Will update on my end.
